pkg/cmd/repo/list/list.go

```diff
@@ -0,0 +1,224 @@
+}
+
+type ListOptions struct {
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+
+	Limit int
+	Owner string
+
+	Visibility string
+	Fork       bool
+	Source     bool
+	Archived   bool
+}
+
+func NewCmdList(f *cmdutil.Factory, runF func(*ListOptions) error) *cobra.Command {
+	opts := ListOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+	}
+
+	var (
+		flagPublic   bool
+		flagPrivate  bool
+		flagSource   bool
+		flagFork     bool
+		flagArchived bool
+	)
+
+	cmd := &cobra.Command{
+		Use:   "list",
+		Args:  cobra.MaximumNArgs(1),
+		Short: "List repositories from a user or organization",
+		RunE: func(c *cobra.Command, args []string) error {
+			if opts.Limit < 1 {
+				return &cmdutil.FlagError{Err: fmt.Errorf("invalid limit: %v", opts.Limit)}
+			}
+
+			if flagPrivate && flagPublic {
+				return &cmdutil.FlagError{Err: fmt.Errorf("specify only one of `--public` or `--private`")}
+			}
+			if flagSource && flagFork {
+				return &cmdutil.FlagError{Err: fmt.Errorf("specify only one of `--source` or `--fork`")}
+			}
+
+			if flagPrivate {
+				opts.Visibility = "private"
+			} else if flagPublic {
+				opts.Visibility = "public"
+			}
+
+			opts.Archived = flagArchived
+			opts.Fork = flagFork
+			opts.Source = flagSource
+
+			if len(args) > 0 {
+				opts.Owner = args[0]
+			}
+
+			if runF != nil {
+				return runF(&opts)
+			}
+			return listRun(&opts)
+		},
+	}
+
+	cmd.Flags().IntVarP(&opts.Limit, "limit", "L", 30, "Maximum number of repositories to list")
+	cmd.Flags().BoolVar(&flagPrivate, "private", false, "Show only private repositories")
+	cmd.Flags().BoolVar(&flagPublic, "public", false, "Show only public repositories")
+	cmd.Flags().BoolVar(&flagSource, "source", false, "Show only source repositories")
+	cmd.Flags().BoolVar(&flagArchived, "archived", false, "Show only archived repositories")
+	cmd.Flags().BoolVar(&flagFork, "fork", false, "Show only forks")
+
+	return cmd
+}
+
+func listRun(opts *ListOptions) error {
+	httpClient, err := opts.HttpClient()
+	if err != nil {
+		return err
+	}
+
+	apiClient := api.NewClientFromHTTP(httpClient)
+
+	isTerminal := opts.IO.IsStdoutTTY()
+
+	owner := opts.Owner
+	if owner == "" {
+		owner, err = api.CurrentLoginName(apiClient, ghinstance.OverridableDefault())
+		if err != nil {
+			return err
+		}
+	}
+
+	filter := FilterOptions{
+		Visibility: opts.Visibility,
+		Fork:       opts.Fork,
+		Source:     opts.Source,
+		Archived:   opts.Archived,
+	}
+
+	listResult, err := listRepos(apiClient, ghinstance.OverridableDefault(), opts.Limit, owner, filter)
+	if err != nil {
+		return err
+	}
+
+	cs := opts.IO.ColorScheme()
+
+	tp := utils.NewTablePrinter(opts.IO)
+
+	notArchived := (filter.Fork || filter.Source) && !filter.Archived
+
+	matchCount := len(listResult.Repositories)
+	now := time.Now()
+
+	for _, repo := range listResult.Repositories {
+		if notArchived && repo.IsArchived {
+			matchCount--
+			listResult.TotalCount--
+			continue
```

Comment:
When doing `gh repo list --fork` or `gh repo list --source`, don't list archived repos.
This mimics the way GitHub does filter repos in the web UI.